lib/location/method_gnss.c

```diff
@@ -0,0 +1,835 @@
+	}
+}
+
+#if defined(CONFIG_NRF_CLOUD_AGPS) || defined(CONFIG_NRF_CLOUD_PGPS)
+static void method_gnss_request_assistance(void)
+{
+	int err = nrf_modem_gnss_read(&agps_request,
+				      sizeof(agps_request),
+				      NRF_MODEM_GNSS_DATA_AGPS_REQ);
+
+	if (err) {
+		LOG_WRN("Reading A-GPS req data from GNSS failed, error: %d", err);
+		return;
+	}
+
+	LOG_DBG("A-GPS request from modem (ephe: 0x%08x alm: 0x%08x flags: 0x%02x)",
+		agps_request.sv_mask_ephe,
+		agps_request.sv_mask_alm,
+		agps_request.data_flags);
+#if defined(CONFIG_LOCATION_METHOD_GNSS_AGPS_EXTERNAL)
+	k_work_submit_to_queue(loc_core_work_queue_get(), &method_gnss_agps_cb_work);
+#else
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+	/* Check the request. If no A-GPS data types except ephemeris or almanac are requested,
+	 * jump to P-GPS (if enabled)
+	 */
+	if (method_gnss_agps_required(&agps_request)) {
+		k_work_submit_to_queue(loc_core_work_queue_get(), &method_gnss_agps_request_work);
+	} else
+#endif
+	{
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+		k_work_submit_to_queue(loc_core_work_queue_get(), &method_gnss_notify_pgps_work);
+#endif
+	}
+#endif // defined(CONFIG_LOCATION_METHOD_GNSS_AGPS_EXTERNAL)
+}
+#endif // defined(CONFIG_NRF_CLOUD_AGPS) || defined(CONFIG_NRF_CLOUD_PGPS)
+
+void method_gnss_event_handler(int event)
+{
+	switch (event) {
+	case NRF_MODEM_GNSS_EVT_PVT:
+		k_work_submit_to_queue(loc_core_work_queue_get(), &method_gnss_pvt_work);
+		break;
+
+	case NRF_MODEM_GNSS_EVT_AGPS_REQ:
+#if defined(CONFIG_NRF_CLOUD_AGPS) || defined(CONFIG_NRF_CLOUD_PGPS)
+		method_gnss_request_assistance();
+#endif
+		break;
+	}
+}
+
+int method_gnss_cancel(void)
+{
+	int err = nrf_modem_gnss_stop();
+	int sleeping;
+	int rrc_idling;
+
+	if (err == NRF_EPERM) {
+		/* Let's convert NRF_EPERM to EPERM to make sure loc_core works properly */
+		err = EPERM;
+		/* NRF_EPERM is logged in loc_core */
+	} else if (err) {
+		LOG_ERR("Failed to stop GNSS");
+	}
+
+	running = false;
+
+	/* Cancel any work that has not been started yet */
+	(void)k_work_cancel(&method_gnss_start_work.work_item);
+
+	/* If we are currently not in PSM, i.e., LTE is running, reset the semaphore to unblock
+	 * method_gnss_positioning_work_fn() and allow the ongoing location request to terminate.
+	 * Otherwise, don't reset the semaphore in order not to lose information about the current
+	 * sleep state.
+	 */
+	sleeping = k_sem_count_get(&entered_psm_mode);
+	if (!sleeping) {
+		k_sem_reset(&entered_psm_mode);
+	}
+
+	/* If we are currently in RRC connected mode, reset the semaphore to unblock
+	 * method_gnss_positioning_work_fn() and allow the ongoing location request to terminate.
+	 * Otherwise, don't reset the semaphore in order not to lose information about the current
+	 * RRC state.
+	 */
+	rrc_idling = k_sem_count_get(&entered_rrc_idle);
+	if (!rrc_idling) {
+		k_sem_reset(&entered_rrc_idle);
+	}
+
+
+	return -err;
+}
+
+#if !defined(CONFIG_NRF_CLOUD_AGPS)
+static bool method_gnss_entered_psm(void)
```

Comment:
I mean the function name as a question :)

Reply:
Yes, but it returns `true` also when device doesn't have PSM enabled, so it's not actually answering that question. :) It could be something like `method_gnss_wait_for_psm_if_needed()`, but not sure if that's very good either.

Reply:
I ended up splitting method_gnss_entered_psm() into two functions.